### PR TITLE
Simplified regex

### LINKED ISSUE #49: [path] Simplified regex aka reordering of hardcoded path to zone mapping
**Is this a BUG REPORT or FEATURE REQUEST?**:
feature

**What happened**:
Currently the regex and the order is hardcoded.

**What you expected to happen**:
Let the user define the order, while still retaining the regex to match yeah directory/subdirectory.
`from=/$1/$3/$2` would map to the _redirect record as: `_redirect.$3.$2.$1.example.com`

### PR DESCRIPTION
**Pull request by erbesharat**
**What this PR does / why we need it**:
Get order from record

**Which issue this PR fixes**:
fixes #49 and #50 

**Special notes for your reviewer**:

**Release note**:

```release-note

```

### CODE REVIEW

File: path.go [hunk 1]
```diff
@@ -4,16 +4,32 @@ import (
...
-func zoneFromPath(host string, path string) (string, int, error) {
+func zoneFromPath(host string, path string, rec record) (string, int, error) {
 	match, err := regexp.Compile("([a-zA-Z0-9]+)")
 	if err != nil {
 		return "", 0, err
 	}
 	pathSlice := match.FindAllString(path, -1)
 	from := len(pathSlice)
+	if rec.From != "" {
+		match, err := regexp.Compile("\\/\\$(\\d+)")
```
[the remark this answers is not in the document]
Comment (stp-ip) [reply path.go]: Might make sense to have these regex ones as a constant.
Right now that would be the default regex + the regex to get the actual reorder hits.

File: path_test.go [hunk 2]
```diff
@@ -8,30 +8,43 @@ func Test_zoneFromPath(t *testing.T) {
...
 			"/",
+			"",
 			"_redirect.example.com",
 			nil,
 		},
+		{
+			"example.com",
+			"/caddy/pkg/v1",
+			"/$1/$3/$2",
+			"_redirect.caddy.v1.pkg.example.com",
```
[the remark this answers is not in the document]
Comment (erbesharat) [reply path_test.go]: I think I've not understood the order correctly.
``example.com/$1/$2/$3/$4``
isn't this the default order?
```
example.com/caddy/pkg/v1
$3 = v1
$2 = pkg
$1 = caddy
```

File: path.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (erbesharat) [reply path.go]: Will assign each regex to a constant var.

File: path_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (stp-ip) [reply path_test.go]: So the default order (without reordering) afaik:
```
example.com/$1/$2/$3
_redirect.$3.$2.$1.example.com
```

[the remark this answers is not in the document]
Comment (stp-ip) [reply path_test.go]: Reasoning: We want to get more specific the deeper the subdomain.
Usually with path it's `/broad/semi-specific/specific`.
With subdomains it's `specific.semi-specific.broad.example.com`

File: path_test.go [hunk 3]
```diff
@@ -33,11 +33,18 @@ func Test_zoneFromPath(t *testing.T) {
 			"_redirect.example.com",
 			nil,
 		},
+		{
+			"example.com",
+			"/caddy/pkg/v1/download",
+			"/$1/$4/$2/$3",
+			"_redirect.v1.pkg.download.caddy.example.com",
```
Comment (stp-ip) [inline path_test.go]: Shouldn't this be: `_redirect.pkg.download.v1.caddy.example.com`?

**Comment by stp-ip** [comment]
LGTM
@alarsyo time for another review?

**Comment by alarsyo** [comment]
Will be reviewed before noon :)

File: path.go [hunk 4]
```diff
@@ -4,16 +4,51 @@ import (
...
 	"regexp"
+	"sort"
+	"strconv"
 	"strings"
 )
 
-func zoneFromPath(host string, path string) (string, int, error) {
-	match, err := regexp.Compile("([a-zA-Z0-9]+)")
+func zoneFromPath(host string, path string, rec record) (string, int, error) {
+	match, err := regexp.Compile("\\/([A-Za-z0-9-._~!$'()*+,;=:@]+)")
```
Comment (alarsyo) [inline path.go]: You forgot to make this a global const :)

File: path.go [hunk 5]
```diff
@@ -4,16 +4,51 @@ import (
...
 	}
-	pathSlice := match.FindAllString(path, -1)
+	pathSubmatchs := match.FindAllStringSubmatch(path, -1)
+	pathSlice := []string{}
+	for _, v := range pathSubmatchs {
+		pathSlice = append(pathSlice, v[1])
+	}
 	from := len(pathSlice)
+	if rec.From != "" {
+		match, err := regexp.Compile("\\/\\$(\\d+)")
```
Comment (alarsyo) [inline path.go]: Same here !

**Comment by stp-ip** [comment]
.Default isn't yet added within the Parser.

**Comment by stp-ip** [comment]
```
if rec.Type == "path" {
  if path == "/" {
    http.Redirect(w, r, rec.Default, rec.Code)
    return nil
    }
  if  path != "" {
    zone, from, _ := zoneFromPath(host, path, rec)
    rec, err = getFinalRecord(zone, from)
      if err != nil {
        return err
        }
    }
  }
```

This should fix the host redirect loop.
Checking or the root path only within type=path

Also when do we redirect to rec.To again? There was a reason we used rec.Default...

**Comment by erbesharat** [comment]
We've used rec.To to redirect when the record type is "host".

**Comment by stp-ip** [comment]
We don't yet seem to default back to _redirect.example.com, if no record is found for the other wildcards.
That means, if `_redirect._._.example` isn't found it will error out instead of using `_redirect.example.com`.
Sidenote, we don't need to check the record again, but can use rec.To from the first iteration of path. Or?

**Comment by stp-ip** [comment]
Also rec.From is not parsed yet leading to a "multiple keys" error.

**Comment by stp-ip** [comment]
Without a rec.Default it end in a redirect loop. It should fallback to rec.To > config default (redirect "") > 404.

**Comment by stp-ip** [comment]
Another bug:
Having a path such as "/one/", but a "from=/$1/$2" fails with index out of range. It shouldn't fail hard. Not sure, what the best failure mode would be so.

example.com/one/ -> _redirect.one.example.com  instead of failing?

File: txtdirect.go [hunk 6]
```diff
@@ -175,16 +179,25 @@ func Redirect(w http.ResponseWriter, r *http.Request, c Config) error {
...
+	fallback, code := getBaseTarget(rec)
 
-	if rec.Type == "path" && path != "" {
-		zone, from, _ := zoneFromPath(host, path)
-		rec, err = getFinalRecord(zone, from)
-		if err != nil {
-			return err
+	if rec.Type == "path" {
+		if path == "/" {
+			http.Redirect(w, r, rec.Default, rec.Code)
```
Comment (stp-ip) [inline txtdirect.go]: If we agree on using rec.Root instead of rec.Default this needs a change here.

Comment (alarsyo) [reply txtdirect.go]: From our Basecamp discussion, I think @stp-ip wanted to rename it, not have both

**Comment by alarsyo** [comment]
Will read through this on my lunch break to give thoughts :)

**Pull request merged**